### PR TITLE
adjust foreman-plugin.mk to build the plugin as a user

### PR DESCRIPTION
**Pull request by evgeni**
learnings from #7068, but instead of modifying each plugin, let's fix plugin.mk and then switch all plugins to it

structure is very much following what we do in the macros for RPM, so don't judge me by the individual steps :laughing:

### CODE REVIEW

**Review by ekohl — COMMENTED** [review]
This feels hard to follow. I wonder if it would make sense to implement something in https://github.com/theforeman/foreman/blob/develop/Rakefile.dist (which already has a `PREFIX` variable.

Another related part: today we symlink `/usr/share/foreman/db` to `/var/lib/foreman/db` and then symlink back some paths. This needed to be done in the past because of sqlite support but that's gone. I wonder how much else there is left that would be considered state. In short: is it possible to keep the `db` directory completely in `/usr/share/foreman`?

**Comment by evgeni** [comment]
> This feels hard to follow.

I know :(
It's mostly copying things from RPM tho.

> I wonder if it would make sense to implement something in https://github.com/theforeman/foreman/blob/develop/Rakefile.dist (which already has a `PREFIX` variable.

Not sure what you want to move there, care to elaborate a bit?

> Another related part: today we symlink `/usr/share/foreman/db` to `/var/lib/foreman/db` and then symlink back some paths. This needed to be done in the past because of sqlite support but that's gone. I wonder how much else there is left that would be considered state. In short: is it possible to keep the `db` directory completely in `/usr/share/foreman`?

It still has the `schema` file, which I think gets updated?

**Comment by ekohl** [comment]
> Not sure what you want to move there, care to elaborate a bit?

Perhaps we can create a new rake task to build a plugin as well. Ideally even use that on both RPM and debs to unify both processes.

> It still has the `schema` file, which I think gets updated?

I think I've also seen some dump file being created. However, perhaps we can move those files and instead create a `db/state` (or something) and symlink that.

**Comment by ehelms** [comment]
[test deb]

**Comment by ehelms** [comment]
I realized, we probably want to test this situation with the new job but it was only checking for changelog changes. So we could fix that with https://github.com/theforeman/jenkins-jobs/pull/104 or we could put a check in to enforce always changing the changelog like we do with RPMs so there is a record of changes.

Thoughts?

**Comment by ekohl** [comment]
I think the latter feels better. In this case it's changing the plugin.mk that ends up being in foreman itself. The only way to rebuild a plugin that uses it is to change the changelog. Alignment with RPM policies is also not a bad thing.

**Comment by evgeni** [comment]
no changelog → big red "no merge for you" sounds good to me.

**Comment by ehelms** [comment]
I think there are two changes for this PR:

 1) Address feedback and add deeper explanation
 2) Bump the changelog so test builds run

**Comment by ehelms** [comment]
Needs a rebase, additionally is there a path forward for this change?

**Comment by evgeni** [comment]
@ekohl @ehelms I've added comments to the code and restructured it a bit, which makes it better readable (IMHO), so how about now? :)

File: debian/bullseye/foreman/foreman-plugin.mk.in [hunk 1]
```diff
@@ -1,16 +1,40 @@
 BUNDLE = @BUNDLE@
 
 build:
-	cp cache/* /usr/share/foreman/vendor/cache/
-	cp $(PLUGIN).rb /usr/share/foreman/bundler.d/
-	cd /usr/share/foreman && ( \
+	# prepare a local foreman installation based on packages
+	mkdir -p ./usr/share
+	cp -r /usr/share/foreman ./usr/share || true
```
Comment (ekohl) [inline debian/bullseye/foreman/foreman-plugin.mk.in:6]: If the whole tree needs to be copied, it does make it a lot slower.

Just thinking out loud: you need the structure to exist and then add something to `bundler.d/` so the plugin exists, right? Could we in the foreman build package add something so you can use an env var to specify the entry and then use it as non-root?

You still may need to copy the vendor directory, or at least its cache. Possibly also NPM's node_modules.

Comment (evgeni) [reply debian/bullseye/foreman/foreman-plugin.mk.in:6]: If by "a lot" you mean "one second", sure ;-)

```
[2022-10-17T08:23:06.186Z] # prepare a local foreman installation based on packages
[2022-10-17T08:23:06.186Z] mkdir -p ./usr/share
[2022-10-17T08:23:06.186Z] cp -r /usr/share/foreman ./usr/share || true
[2022-10-17T08:23:07.287Z] # clean up possibly existing state we copied
```

(from https://ci.theforeman.org/blue/organizations/jenkins/foreman-packaging-deb-pr-test/detail/foreman-packaging-deb-pr-test/813/pipeline/66 which ran with these changes)

And no, I need the whole code to be reachable, so that `bundle exec rake assets:precompile` works, which requires the app to be bootable.

Comment (ekohl) [reply debian/bullseye/foreman/foreman-plugin.mk.in:6]: Hmm, I guess I was still thinking about non-NVME storage. Then this is easier.

File: debian/bullseye/foreman/foreman-plugin.mk.in [hunk 2]
```diff
@@ -1,16 +1,40 @@
...
+
+	# add plugin code to the local foreman installation
+	cp cache/* ./usr/share/foreman/vendor/cache/
+	cp $(PLUGIN).rb ./usr/share/foreman/bundler.d/
+
+	# perform the actual build of the plugin
+	cd ./usr/share/foreman && ( \
 		/usr/bin/foreman-ruby $(BUNDLE) install --local && \
 		if [ "$(PLUGIN_HAS_PACKAGE_JSON)" = "true" ]; then /usr/bin/npm install --no-audit --no-optional --unsafe-perm; fi && \
+		/usr/bin/foreman-ruby $(BUNDLE) exec rake security:generate_encryption_key && \
```
Comment (ekohl) [inline debian/bullseye/foreman/foreman-plugin.mk.in:27]: What's the reasoning behind removing the old key and then generating a new one?

File: debian/bullseye/foreman/foreman-plugin.mk.in [hunk 3]
```diff
@@ -1,16 +1,40 @@
...
-	cd /usr/share/foreman && ( \
+	# prepare a local foreman installation based on packages
+	mkdir -p ./usr/share
+	cp -r /usr/share/foreman ./usr/share || true
+
+	# clean up possibly existing state we copied
+	mkdir -p ./var/lib/foreman
+	cp -r /var/lib/foreman/db ./var/lib/foreman || true
+	unlink ./usr/share/foreman/db
+	mkdir ./usr/share/foreman/db
```
Comment (ekohl) [inline debian/bullseye/foreman/foreman-plugin.mk.in:12]: Unrelated, but now that we've dropped SQLite I wonder if we still have anything that we write to at runtime in `db/`. Migrating `/var/lib/foreman/db` back to `/usr/share/foreman/db` is probably painful so I've never bothered but still something I wonder about every now and then.

File: debian/bullseye/foreman/foreman-plugin.mk.in (hunk 2 above)
Comment (evgeni) [reply debian/bullseye/foreman/foreman-plugin.mk.in:27]: Ah, most probably:

```console
# ls -alh /usr/share/foreman/config/initializers/encryption_key.rb 
lrwxrwxrwx. 1 root root 30 Jun 22 08:02 /usr/share/foreman/config/initializers/encryption_key.rb -> /etc/foreman/encryption_key.rb
```

(A user won't be able to read that)

**Review by ekohl — APPROVED** [review]
I think we can try this. Still enough time until we get to 3.5 branching.
